### PR TITLE
IINT-7229: add InsightAppSec Integration

### PR DESCRIPTION
**Pull request by gastonyelmini**
# Description

Thank you for contributing to a JupiterOne integration!

Please include a summary of the change and which issue is fixed. Please also
include relevant motivation and context. List any dependencies that are required
for this change.

## Type of change

Please leave any irrelevant options unchecked.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)
- [ ] This change requires a documentation update

### General Development Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### Integration Development Checklist:

Please leave any irrelevant options unchecked.

- [ ] I have checked for additional permissions required to call any new API
      endpoints, and have documented any additional permissions in
      `jupiterone.md`, where necessary.
- [ ] My changes properly paginate the target service provider's API
- [ ] My changes properly handle rate limiting of the target service provider's
      API
- [ ] My new integration step is instrumented to execute in the correct order
      using `dependsOn`
- [ ] I have referred to the
      [JupiterOne data model](https://github.com/JupiterOne/data-model/tree/main/src/schemas)
      to ensure that any new entities/relationships, and relevant properties,
      match the recommended model for this class of data
- [ ] I have updated the `CHANGELOG.md` file to describe my changes
- [ ] When changes include modifications to existing graph data ingestion, I've
      reviewed all existing managed questions referencing the entities,
      relationships, and their property names, to ensure those questions still
      function with my changes.

### CODE REVIEW

File: src/steps/app-sec/steps/engines/engine.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+  if (defaultEngineGroupEntity) {
+    await jobState.addRelationship(
+      createDirectRelationship({
+        _class: RelationshipClass.HAS,
+        from: defaultEngineGroupEntity,
+        to: cloudDefaultEngineEntity,
+      }),
+    );
+  }
```
Comment (RonaldEAM) [inline src/steps/app-sec/steps/engines/engine.ts:49]: can be changed to not use findEntity

File: src/steps/app-sec/steps/engines/engine.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            await jobState.addRelationship(
+              createDirectRelationship({
+                _class: RelationshipClass.HAS,
+                from: engineGroupEntity,
+                to: engineEntity,
+              }),
+            );
+          }
+
+          await jobState.addEntity(engineEntity);
```
Comment (RonaldEAM) [inline src/steps/app-sec/steps/engines/engine.ts:70]: shouldn't this be placed before the addRelationship? I'm not sure if that matters but it's clear on the logic

File: src/steps/app-sec/steps/scan-configs/scan-configs.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  context: IntegrationStepExecutionContext<IntegrationConfig>,
+): Promise<void> {
+  const {
+    jobState,
+    instance: { config },
+    logger,
+  } = context;
+  const client = new Rapid7InsighAppSecClient(config, logger);
+
+  const defaultCloudEngineGroup = await jobState.findEntity(
```
Comment (RonaldEAM) [inline src/steps/app-sec/steps/scan-configs/scan-configs.ts:29]: findEntity is not necessary, hasKey can be used instead where it's needed

File: src/steps/app-sec/steps/scan-configs/scan-configs.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  );
+
+  await client.seachAndIterateResources<ScanConfig>(
+    async (scanConfig) => {
+      const scanConfigEntity = createScanConfigEntity(scanConfig);
+
+      let engineGroupEntityToAssing: Entity | null = defaultCloudEngineGroup;
+
+      if (scanConfig.assignment?.id) {
+        const engineGroupEntity = await jobState.findEntity(
```
Comment (RonaldEAM) [inline src/steps/app-sec/steps/scan-configs/scan-configs.ts:40]: findEntity is not necessary, hasKey can be used instead where it's needed

File: src/steps/app-sec/steps/scan-configs/scan-configs.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+  const defaultCloudEngineGroup = await jobState.findEntity(
+    getEngineGroupKey(DEFAULT_CLOUD_ENGINE_GROUP_ID),
+  );
+
+  await client.seachAndIterateResources<ScanConfig>(
+    async (scanConfig) => {
+      const scanConfigEntity = createScanConfigEntity(scanConfig);
+
+      let engineGroupEntityToAssing: Entity | null = defaultCloudEngineGroup;
```
Comment (RonaldEAM) [inline src/steps/app-sec/steps/scan-configs/scan-configs.ts:37]: you have a typo here on "assign"

File: src/steps/app-sec/steps/scans/scans.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    async (scanConfigEntity) => {
+      const scanConfigRawData = getRawData<ScanConfig>(scanConfigEntity);
+
+      await client.seachAndIterateResources<Scan>(
+        async (scan) => {
+          const scanEntity = creteScanEntity(scan);
+
+          // Create Scan -> PROTECTS -> App
+          if (scan.app?.id) {
+            const appEntity = await jobState.findEntity(
```
Comment (RonaldEAM) [inline src/steps/app-sec/steps/scans/scans.ts:39]: findEntity is not necessary here

File: src/steps/insightvm/steps/sites/sites.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  context: IntegrationStepExecutionContext<IntegrationConfig>,
+): Promise<void> {
+  const { jobState } = context;
+
+  await jobState.iterateEntities(
+    {
+      _type: Rapid7InsightVMEntities.INSIGHT_VM_SITE._type,
+    },
+    async (site) => {
+      const productEntity = await jobState.findEntity(getProductKey('IVM'));
```
Comment (RonaldEAM) [inline src/steps/insightvm/steps/sites/sites.ts:43]: findEntity is not necessary here

File: src/client.ts [hunk 8]
```diff
@@ -74,16 +85,22 @@ export class APIClient {
   public async verifyAuthentication(): Promise<void> {
     try {
       await this.executeAPIRequestWithRetries(
-        `${this.integrationConfig.apiUrl}/validate`,
+        `https://us.${API_BASE_URL}/validate`,
```
Comment (RonaldEAM) [inline src/client.ts:88]: is it always going to be `us` here?

File: src/client.ts [hunk 9]
```diff
@@ -74,16 +85,22 @@ export class APIClient {
...
-        code: err.response.status,
+        message: `Unable to verify credentials: ${err.status} ${err.statusText}`,
+        code: err.statusText,
       });
     }
   }
 
+  public async fetchAccountProducts(): Promise<Product[]> {
+    return this.executeAPIRequestWithRetries<Product[]>(
+      `https://us.${API_BASE_URL}/account/api/1/products`,
```
Comment (RonaldEAM) [inline src/client.ts:100]: is it always going to be `us` here?

File: src/client.ts (hunk 8 above)
Comment (gastonyelmini) [reply src/client.ts:88]: Yep it is

**Review by RonaldEAM — APPROVED** [review]

**Pull request merged**